common/thorhelper/thorread.cpp

```diff
@@ -0,0 +1,825 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2012 HPCC Systems®.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#include "jliball.hpp"
+
+#include "thorfile.hpp"
+
+#include "eclhelper.hpp"
+#include "eclrtl.hpp"
+#include "eclrtl_imp.hpp"
+#include "rtlfield.hpp"
+#include "rtlds_imp.hpp"
+#include "rtldynfield.hpp"
+
+#include "rmtclient.hpp"
+#include "rmtfile.hpp"
+
+#include "thorread.hpp"
+#include "rtlcommon.hpp"
+#include "thorcommon.hpp"
+
+//---------------------------------------------------------------------------------------------------------------------
+
+static Owned<IPropertyTree> emptyTree = createPTree();
+
+//---------------------------------------------------------------------------------------------------------------------
+
+constexpr size32_t defaultReadBufferSize = 0x10000;
+
+//Cannot use InterfaceOf<> because multiple base interfaces are derived from IInterface
+class DiskRowReader : implements IAllocRowStream, implements IRawRowStream, implements IDiskRowReader, implements IThorDiskCallback, public CInterface
+{
+public:
+    DiskRowReader(unsigned _expectedCrc, IOutputMetaData & _expected, unsigned _projectedCrc, IOutputMetaData & _projected, unsigned _actualCrc, IOutputMetaData & _actual, IPropertyTree * options);
+    IMPLEMENT_IINTERFACE
+
+    virtual IRawRowStream * queryRawRowStream() override;
+    virtual IAllocRowStream * queryAllocatedRowStream(IEngineRowAllocator * _outputAllocator) override;
+
+    virtual bool getCursor(MemoryBuffer & cursor) override;
```

Comment:
can be const method?

Reply:
the current implementation calls a non const function.  Something to return to, but not change at the moment.